packages/wallets/src/evm/connectors/smart-wallet/lib/account.ts

```diff
@@ -143,4 +141,49 @@ export class AccountAPI extends BaseAccountAPI {
   async isAcountDeployed() {
     return !(await this.checkAccountPhantom());
   }
+
+  async isAccountApproved() {
+    if (!this.params.erc20PaymasterAddress || !this.params.erc20TokenAddress) {
+      return true;
+    }
+
+    const swAddress = await this.getCounterFactualAddress();
+    const erc20Token = await this.sdk.getContract(
+      this.params.erc20TokenAddress,
+    );
+
+    const allowance = await erc20Token.call("allowance", [
+      swAddress,
+      this.params.erc20PaymasterAddress,
+    ]);
+
+    return allowance.gte(BigNumber.from(2).pow(96).sub(1));
+  }
+
+  async createApproveTx(): Promise<providers.TransactionRequest | undefined> {
+    if (await this.isAccountApproved()) {
+      return undefined;
+    }
+
+    const amountToApprove = BigNumber.from(2).pow(96).sub(1);
+    const ethersSigner = new ethers.Wallet(LOCAL_NODE_PKEY, this.provider);
+    const erc20Contract = new Contract(
+      this.params.erc20TokenAddress as string,
+      [
+        "function approve(address spender, uint256 amount) public returns (bool)",
```

Comment:
you shouldnt have to use ethers here, just use getContract and contract.erc20.approve.prepare()

Reply:
that'll estimate and cause problems i want to avoid i believe

Reply:
in that case you can use `contract.encoder.encodeFunction(....)` - but yeah not a big deal